commec/json_io.py

```diff
@@ -0,0 +1,309 @@
+    ''' Converts a ScreenData class object into a JSON file at the given filepath.'''
+    with open(output_json_filepath, "w", encoding="utf-8") as json_file:
+        json.dump(asdict(input_screendata), json_file, indent=4)
+
+def encode_dict_to_screen_data(input_dict : dict):
+    ''' Converts a dictionary into a ScreenData object, 
+    any keys within the dictionary not part of the ScreenData format are lost.
+    any missing information will be simple set as defaults.'''
+    return dict_to_dataclass(ScreenData, input_dict)
+
+
+
+# Convert the dictionary back to the dataclass or list of dataclass
+def dict_to_dataclass(cls: Type, data: Dict[str, Any]) -> Any:
+    ''' 
+    Convert a dict, into appropriate dataclass, or list of dataclass, 
+    invalid keys to the dataclass structure are ignored.
+    '''
+    # Prepare a dictionary for filtered data
+    filtered_data = {}
+
+    for f in fields(cls):
+        field_name = f.name
+        field_type = f.type
+        if field_name in data:
+            field_value = data[field_name]
+
+            # Check if the field is a dataclass
+            if is_dataclass(field_type):
+                filtered_data[field_name] = dict_to_dataclass(field_type, field_value)
+                continue
+
+            # Check if the field is a list of dataclasses
+            if get_origin(field_type) is list:
+                item_type = get_args(field_type)[0]
+                if is_dataclass(item_type) and isinstance(field_value, list):
+                    filtered_data[field_name] = [
+                        dict_to_dataclass(item_type, item) for item in field_value
+                        if isinstance(item, dict) 
+                            and any(key in {f.name for f in fields(item_type)} for key in item.keys())
+                            or isinstance(item, item_type)]
+                    continue
+                filtered_data[field_name] = field_value
+                continue
+
+            # Handle other field types
+            filtered_data[field_name] = field_value
+
+    # Create an instance of the dataclass with the filtered data
+    return cls(**filtered_data)
+
+def get_screen_data_from_json(input_json_filepath: string):
+    ''' Loads a JSON file from given filepath and returns a populated ScreenData object from its contents.'''
+    json_string : str
+    with open(input_json_filepath, "r", encoding="utf-8") as json_file:
+        # Read the file contents as a string
+        json_string = json_file.read()
+    my_data : dict = json.loads(json_string)
+
+    # Check version of imported json.
+    input_version = my_data["commec_info"]["json_output_version"]
+    if not input_version == JSON_COMMEC_FORMAT_VERSION:
+        raise RuntimeError("""Version difference between input (v.{input_version}) and
+                            expected (v.{JSON_COMMEC_FORMAT_VERSION}) JSON state file: {input_json_filepath}""")
+    return encode_dict_to_screen_data(my_data)
+
+def test1():
```

Comment:
Main piece of feedback is that I'd love to move these into a test_json_iso.py class under /tests, rather than having them directly in this module.